app/assets/javascripts/discourse/app/mixins/uppy-upload.js

```diff
@@ -0,0 +1,177 @@
+
+    if (!this.id) {
+      warn(
+        "uppy needs a unique id, pass one in to the component implementing this mixin",
+        {
+          id: "discourse.upload.missing-id",
+        }
+      );
+    }
+
+    this.set(
+      "uppyInstance",
+      new Uppy({
+        id: this.id,
+        autoProceed: this.autoStartUploads,
+
+        // need to use upload_type because uppy overrides type with the
+        // actual file type
+        meta: deepMerge({ upload_type: this.type }, this.data || {}),
+
+        onBeforeFileAdded: (currentFile) => {
+          const validationOpts = deepMerge(
+            {
+              bypassNewUserRestriction: true,
+              user: this.currentUser,
+              siteSettings: this.siteSettings,
+            },
+            this.validateUploadedFilesOptions()
+          );
+          const isValid = validateUploadedFile(currentFile, validationOpts);
+          this.setProperties({ uploadProgress: 0, uploading: isValid });
+          return isValid;
+        },
+
+        onBeforeUpload: (files) => {
+          let tooMany = false;
+          const fileCount = Object.keys(files).length;
+          const maxFiles = this.getWithDefault(
+            "maxFiles",
+            this.siteSettings.simultaneous_uploads
+          );
+
+          if (this.allowMultipleFiles) {
+            tooMany = maxFiles > 0 && fileCount > maxFiles;
+          } else {
+            tooMany = fileCount > 1;
+          }
+
+          if (tooMany) {
+            bootbox.alert(
+              I18n.t("post.errors.too_many_dragged_and_dropped_files", {
+                count: this.allowMultipleFiles ? maxFiles : 1,
+              })
+            );
+            this._reset();
+            return false;
+          }
+        },
+      })
+    );
+
+    this.uppyInstance.use(DropTarget, { target: this.element });
+
+    this.uppyInstance.on("progress", (progress) => {
+      this.set("uploadProgress", progress);
+    });
+
+    this.uppyInstance.on("upload-success", (_file, response) => {
+      this.uploadDone(response.body);
+      this._reset();
+    });
+
+    this.uppyInstance.on("upload-error", (file, error, response) => {
+      displayErrorForUppyUpload(response, file.name, this.siteSettings);
+      this._reset();
+    });
+
+    // later we will use the uppy direct s3 uploader based on enable_s3_uploads,
+    // for now we always just use XHR uploads
+    this._useXHRUploads();
+  },
+
+  _useXHRUploads() {
+    this.uppyInstance.use(XHRUpload, {
+      endpoint: this._xhrUploadUrl(),
+      headers: {
+        "X-CSRF-Token": this.session.get("csrfToken"),
+      },
+    });
+  },
+
+  _xhrUploadUrl() {
+    return (
+      getUrl(this.getWithDefault("uploadUrl", "/uploads")) +
+      ".json?client_id=" +
+      (this.messageBus && this.messageBus.clientId) +
+      this.uploadUrlParams
+    );
+  },
+
+  _bindFileInputChangeListener() {
+    this.fileInputEl.addEventListener("change", (event) => {
```

Comment:
jQuery file uploader does this magically, we need to handle it for uppy (unless we want the whole UI dashboard ultra loot box experience that you can get from uppy)